src/track/macro.h

```diff
@@ -0,0 +1,78 @@
+#pragma once
+
+#include <QtCore>
+#include <memory>
+
+#include "engine/engine.h"
+#include "proto/macro.pb.h"
+#include "track/macroaction.h"
+#include "util/db/dbid.h"
+namespace proto = mixxx::track::io;
+
+const QLoggingCategory macroLoggingCategory("macros");
+constexpr int kMacrosPerChannel = 16;
+
+/// A Macro stores a list of MacroActions as well as its current state and label.
+class Macro {
+  public:
+    static QByteArray serialize(const QList<MacroAction>& actions);
+    static QList<MacroAction> deserialize(const QByteArray& serialized);
+
+    enum class StateFlag {
+        Enabled = 1u,
+        Looped = 2u,
+    };
+    Q_DECLARE_FLAGS(State, StateFlag);
+
+    explicit Macro(const QList<MacroAction>& actions = {},
+            const QString& label = "",
+            State state = State(StateFlag::Enabled),
+            int dbId = DbId::s_invalidValue);
+
+    bool isDirty() const;
+    int getId() const;
+
+    QString getLabel() const;
+    void setLabel(const QString&);
+
+    bool isEnabled() const;
+    bool isLooped() const;
+    const State& getState() const;
+    void setState(StateFlag flag, bool enable = true);
+
+    bool isEmpty() const;
+    unsigned int size() const;
+
+    const QList<MacroAction>& getActions() const;
+    void addAction(const MacroAction& action);
+
+    mixxx::audio::FramePos getStart() const;
+    /// Sets the end of the Macro (relevant for looping).
+    void setEnd(mixxx::audio::FramePos framePos);
+
+    void clear();
+
+  private:
+    void setDirty(bool dirty = true);
+    void setId(int id);
+
+    bool m_bDirty;
```

Comment:
Is this dirty flag really needed? I don't recommend to introduce stateful, hidden behavior unless it is actually needed.